src/net.cpp

```diff
@@ -943,35 +981,95 @@ void ThreadSocketHandler()
+                        }
+
+                        // Protect connections with certain characteristics
+
+                        std::sort(vEvictionCandidates.begin(), vEvictionCandidates.end(), ReverseCompareNodeTimeConnected);
+                        vEvictionCandidates.erase(vEvictionCandidates.end() - std::min(64, static_cast<int>(vEvictionCandidates.size())), vEvictionCandidates.end());
+
+                        CompareNetGroupKeyed comparerNetGroupKeyed(vSecretKey);
+                        std::sort(vEvictionCandidates.begin(), vEvictionCandidates.end(), comparerNetGroupKeyed);
+                        vEvictionCandidates.erase(vEvictionCandidates.end() - std::min(4, static_cast<int>(vEvictionCandidates.size())), vEvictionCandidates.end());
+
+                        std::sort(vEvictionCandidates.begin(), vEvictionCandidates.end(), ReverseCompareNodeMinPingTime);
+                        vEvictionCandidates.erase(vEvictionCandidates.end() - std::min(8, static_cast<int>(vEvictionCandidates.size())), vEvictionCandidates.end());
+
+                        if (!vEvictionCandidates.empty()) {
+                            // Identify CNetAddr with the most connections
+                            CNetAddr naMostConnections;
+                            unsigned int nMostConnections = 0;
+                            std::map<CNetAddr, std::vector<CNode*> > mapAddrCounts;
+                            BOOST_FOREACH(CNode *node, vEvictionCandidates) {
+                                mapAddrCounts[node->addr].push_back(node);
+
+                                if (mapAddrCounts[node->addr].size() > nMostConnections) {
+                                    nMostConnections = mapAddrCounts[node->addr].size();
+                                    naMostConnections = node->addr;
+                                }
+                            }
+
+                            vEvictionCandidates = mapAddrCounts[naMostConnections];
+
+                            // Disconnect the most recent connection from the CNetAddr with the most connections
+                            std::sort(vEvictionCandidates.begin(), vEvictionCandidates.end(), ReverseCompareNodeTimeConnected);
+                            vEvictionCandidates[0]->fDisconnect = true;
+
+                            nInbound--;
+                        }
+                    }
+
+                    if (nInbound >= nMaxConnections - MAX_OUTBOUND_CONNECTIONS) {
+                        LogPrint("net", "no peers to evict, disconnecting new client\n");
+                        pnode->fDisconnect = true;
+                    }
                 }
-                else if (CNode::IsBanned(addr) && !whitelisted)
-                {
-                    LogPrintf("connection from %s dropped (banned)\n", addr.ToString());
+
+                if (pnode->fDisconnect) {
                     CloseSocket(hSocket);
+                    delete pnode;
```

Comment:
Bah.
I wonder if we could use e.g. boost::scoped_ptr to avoid this. It's too easy to get memory leaks with manual deallocation along a subset of code paths, especially when factoring in exceptions.